instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/CoreRerouteTest.kt

```diff
@@ -123,36 +138,83 @@ class CoreRerouteTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.jav
                         reasons: List<RouterFailure>,
                         routeOptions: RouteOptions
                     ) {
-                        // no impl
+                        Log.e("DEBUG", "onFailure reasons=$reasons")
                     }
 
                     override fun onCanceled(
                         routeOptions: RouteOptions,
                         routerOrigin: RouterOrigin
                     ) {
-                        // no impl
+                        Log.e("DEBUG", "onCanceled")
                     }
                 }
             )
         }
 
         // wait for tracking to start
-        Espresso.onIdle()
+        try {
+            Espresso.onIdle()
+        } catch (ex: RuntimeException) {
```

Comment:
Which `RuntimeException` is being thrown here? Curious why we expect it 🤔 . Is this to help us catch issues when they arise?

Reply:
We could add this check to the `MapboxHistoryTestRule`

for example. add a function

``` kotlin
mapboxHistoryTestRule.stopRecordingOnCrash {
    Espresso.onIdle()
}
```

``` kotlin
MapboxHistoryTestRule {
    fun stopRecordingOnCrash(runner: () -> Unit) {
        try {
            runner()
        } catch (t: Throwable) {
            val countDownLatch = CountDownLatch(1)
            mapboxNavigation.historyRecorder.stopRecording {
                Log.e("DEBUG", "no location tracking history path=$it")
                countDownLatch.countDown()
            }
            countDownLatch.await()
        }
    }
}

Reply:
> Which `RuntimeException` is being thrown here? Curious why we expect it 🤔 . Is this to help us catch issues when they arise?

```
java.lang.RuntimeException: java.util.concurrent.ExecutionException: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteProgressStateIdlingResource] to become idle timed out
	at androidx.test.espresso.Espresso.onIdle(Espresso.java:295)
	at androidx.test.espresso.Espresso.onIdle(Espresso.java:314)
	at com.mapbox.navigation.instrumentation_tests.core.MapboxHistoryTest.verify_history_files_are_recorded_and_readable(MapboxHistoryTest.kt:142)
	at java.lang.reflect.Method.invoke(Native Method)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at androidx.test.internal.runner.junit4.statement.RunBefores.evaluate(RunBefores.java:80)
	at androidx.test.internal.runner.junit4.statement.RunAfters.evaluate(RunAfters.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at androidx.test.rule.GrantPermissionRule$RequestPermissionStatement.evaluate(GrantPermissionRule.java:134)
	at androidx.test.rule.ActivityTestRule$ActivityStatement.evaluate(ActivityTestRule.java:531)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at com.schibsted.spain.barista.rule.cleardata.ClearDatabaseRule$apply$1.evaluate(ClearDatabaseRule.kt:30)
	at com.schibsted.spain.barista.rule.cleardata.ClearPreferencesRule$1.evaluate(ClearPreferencesRule.java:26)
	at com.schibsted.spain.barista.rule.cleardata.ClearFilesRule$apply$1.evaluate(ClearFilesRule.kt:25)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at androidx.test.internal.runner.TestExecutor.execute(TestExecutor.java:56)
	at androidx.test.runner.AndroidJUnitRunner.onStart(AndroidJUnitRunner.java:395)
	at android.app.Instrumentation$InstrumentationThread.run(Instrumentation.java:1879)
Caused by: java.util.concurrent.ExecutionException: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteProgressStateIdlingResource] to become idle timed out
	at java.util.concurrent.FutureTask.report(FutureTask.java:94)
	at java.util.concurrent.FutureTask.get(FutureTask.java:164)
	at androidx.test.espresso.Espresso.onIdle(Espresso.java:287)
	... 46 more
Caused by: androidx.test.espresso.IdlingResourceTimeoutException: Wait for [RouteProgressStateIdlingResource] to become idle timed out
	at androidx.test.espresso.IdlingPolicy.handleTimeout(IdlingPolicy.java:61)
	at androidx.test.espresso.base.UiControllerImpl$5.resourcesHaveTimedOut(UiControllerImpl.java:428)
	at androidx.test.espresso.base.IdlingResourceRegistry$Dispatcher.handleTimeout(IdlingResourceRegistry.java:478)
	at androidx.test.espresso.base.IdlingResourceRegistry$Dispatcher.handleMessage(IdlingResourceRegistry.java:411)
	at android.os.Handler.dispatchMessage(Handler.java:98)
	at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:148)
	at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:519)
	at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:441)
	at androidx.test.espresso.Espresso$1.run(Espresso.java:277)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:423)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at android.os.Handler.handleCallback(Handler.java:739)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5417)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
```

And yeah the idea is to capture and retrieve the history files generated up that point, so we can troubleshoot any potential regressions easily. We should add actually history files support to all of our Instrumentation tests. Will cut a follow up ticket.

Reply:
> Will cut a follow up ticket.

https://github.com/mapbox/mapbox-navigation-android/issues/4927